libass/ass_fontselect.c

```diff
@@ -185,6 +190,80 @@ static void load_fonts_from_dir(ASS_Library *library, const char *dir)
         }
     }
     closedir(d);
+#else
+    // The dir name must ending with \*, D:\temp not work, D:\temp\* works well.
+    size_t dir_len = strlen(dir);
+    char* tdir = malloc(dir_len + 3);
+    strcpy(tdir, dir);
+    strcat(tdir, "\\*");
+    WIN32_FIND_DATAW wfdata;
+    HANDLE whFind;
+    wchar_t *wdir;
+    int wlen;
+    
+    wlen = MultiByteToWideChar(CP_UTF8, 0, tdir, -1, NULL, 0);
+    if (wlen > 0) {
+        wdir = (wchar_t*) malloc(sizeof(wchar_t) * wlen);
+        if (MultiByteToWideChar(CP_UTF8, 0, tdir, -1, wdir, wlen)) {
+            whFind = FindFirstFileW(wdir, &wfdata);
+            if (whFind != INVALID_HANDLE_VALUE) {
+                do {
+                    char *filename;
+                    int fn_len;
+                    fn_len = WideCharToMultiByte(CP_UTF8, 0, wfdata.cFileName, -1, NULL, 0, NULL, FALSE);
+                    if (fn_len > 0) {
+                        filename = (char*) malloc(fn_len);
+                        if (WideCharToMultiByte(CP_UTF8, 0, wfdata.cFileName, -1, filename, fn_len, NULL, FALSE)) {
+                            if (!strcmp(filename, ".") || !strcmp(filename, "..")) {
+                                free(filename);
+                                continue;
+                            }
+                            size_t buffsize = strlen(filename) + strlen(tdir) + 2;
+                            char *fullname = (char*) malloc(buffsize);
+                            snprintf(fullname, buffsize, "%s/%s", dir, filename);
+                            size_t bufsize = 0;
+                            ass_msg(library, MSGL_INFO, "Loading font file '%s'", fullname);
+                            void *data = read_file(library, fullname, &bufsize);
+                            if (data) {
+                                ass_add_font(library, filename, data, bufsize);
+                                free(data);
+                            }
+                            free(fullname);
+                        }
+                        free(filename);
+                    }
+                } while (FindNextFileW(whFind, &wfdata) != 0);
+                FindClose(whFind);
+                free(tdir);
+                free(wdir);
+                return;
+            }
+        }
+        free(wdir);
+    }
+    WIN32_FIND_DATAA fdata;
+    HANDLE hFind;
+    hFind = FindFirstFileA(tdir, &fdata);
```

Comment:
I suspect this will be prone to the exact same errors as the `readdir` implementation, i.e. not all files in the dir can be read with some just not being listed at all and others being listed *(with a fallback name?)* but not being able to be opened, see #500.
Ideally, if UTF-8 conversion or opening fails, the path should be tried to be converted *from ANSI* to MS' “wide char” and tried to be opened via the same “wide char” version as above. This would also allow for less code duplication in the Microsoft-API-path.

Qutoing a relevant section of astiob's comment from #500:
> `         return MultiByteToWideChar(CP_ACP, 0, fname_raw, -1,`
>
> The file path encoding can be CP_OEM, not only CP_ACP, depending on the current value of `AreFileApisANSI`.
>
> There’s also this comment in FFmpeg that says CP_ACP conversions are forbidden in Windows Store. And judging by VLC’s patches to libass, VLC is (or wants to be) in the Windows Store, so they might appreciate us not requiring them to add yet another patch. I haven’t been able to find other sources corroborating this comment’s claim, however.

With the mentioned FFmpeg comment reading:
> […], but doing explicit conversions with CP_ACP is allegedly forbidden in windows store apps (or windows phone), […]

Which also doesn't sound to confident in this restriction actually being a thing. Do you know more about this?
If `CP_ACP` works everywhere you'll "only" have to consider `AreFileApisANSI` for the conversion from "ANSI/OEM", otherwise more ifdeffery it is \o/

Reply:
I can implement a new function with a CodePage parameter. then call that function in `load_fonts_from_dir`

Reply:
Now I put CP_OEMCP before CP_ACP

Reply:
That's not good, it always tries both OEMCP and ACP and potenttially loads multiple dirs  instead of only the inteded one as *all* encodings are *always* tried.
Ignoring the alleged WinRT restriction for now, here's what you probably want to do according to my understanding:

 1) try decoding as UTF-8
     - if fails, goto `2)`
     - if succesful try to open with `FindFirstFileW`
       - if fails goto `2)`
       - if succesful goto `4)`
  2) check [`AreFileApisANSI()`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-arefileapisansi),
      if true use `CP_ACP` in `3)` for `cur_cp`, otherwise use `CPOEMCP`
  3) try decoding as `cur_cp`
    - if fails abort
    - if succesful try opening with `FindFirstFileW`
       - if fails abort
  4) traverse the opened directory with `FindNextFileW`